### PR TITLE
[Feature] Sparse-sparse matrix multiplication, addition, and masking

### PR DESCRIPTION
**Pull request by BarclayII**
## Description
This is the C++ implementation of the RFC #2718 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by BarclayII** [comment]
CUDA implementations and Python integration will follow up in later PRs.

**Comment by BarclayII** [comment]
Still needs further optimization as the current implementation is slower than scipy's.

**Comment by BarclayII** [comment]
Done some benchmarking on my own laptop (8 cores) with the three functions against scipy.

CSRMM that computes A * B (A is of size `MxN` and B is of size `NxP`):

|M|N|P|nnzA|nnzB|scipy csrmm (s)|own csrmm (s)|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|40|50|60|200|300|0.0020560999982990324|0.0020250999987183604|
|4000|5000|6000|20000|30000|0.014718999998876825|0.02095310000004247|
|100000|200000|250000|800000|1500000|4.1519647000022815|1.7335296999990533|

CSR Sum that computes `A + B`  where A and B have the same size:

|M|N|nnz|scipy csrsum|own csrsum|
|:-:|:-:|:-:|:-:|:-:|
|40|50|200|0.00124379999761004|0.0037225999985821545|
|4000|5000|20000|0.004637699999875622|0.01067499999771826|
|100000|200000|800000|0.12096250000104192|0.24896369999987655|

CSR mask that computes `A[B != 0]` where A and B have the same size:

|M|N|nnz|scipy csrmask|own csrmask|
|:-:|:-:|:-:|:-:|:-:|
|40|50|200|0.0037560000018856954|0.0007003999999142252|
|4000|5000|20000|0.008930500000133179|0.0029613000006065704|
|100000|200000|800000|0.24510670000017853|0.06558020000011311|

File: src/array/kernel_decl.h [hunk 1]
```diff
@@ -92,6 +93,32 @@ void BackwardSegmentCmp(NDArray feat,
...
+    const CSRMatrix& A,
+    NDArray A_weights,
+    const CSRMatrix& B,
+    NDArray B_weights);
+
+/*!
+ * \brief Sparse-sparse matrix summation.
+ */
+template <int XPU, typename IdType, typename DType>
+std::pair<CSRMatrix, NDArray> CSRSum(
```
Comment (yzh119) [inline src/array/kernel_decl.h:112]: I think `CSRAdd` is more appropriate.

File: src/array/cpu/csr_mm.cc [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  C_indptr_data[M] = nnz;
+
+  IdArray C_indices = IdArray::Empty({nnz}, A.indices->dtype, A.indices->ctx);
+  NDArray C_weights = NDArray::Empty({nnz}, A_weights->dtype, A_weights->ctx);
+  IdType* C_indices_data = C_indices.Ptr<IdType>();
+  DType* C_weights_data = C_weights.Ptr<DType>();
+
+  // Compute indices and data in parallel
+  phmap::flat_hash_map<IdType, DType> map;
+#pragma omp parallel for firstprivate(map)
```
Comment (yzh119) [inline src/array/cpu/csr_mm.cc:70]: Does `firstprivate` mean `map` would be threadlocal?

File: src/array/cpu/csr_sum.cc [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  for (IdType i = 0; i < M; ++i) {
+    set.clear();
+    for (int64_t k = 0; k < n; ++k) {
+      for (IdType u = A_indptr[k][i]; u < A_indptr[k][i + 1]; ++u)
+        set.insert(A_indices[k][u]);
+    }
+    C_indptr_data[i] = set.size();
+  }
+
+  IdType len = 0;
```
Comment (yzh119) [inline src/array/cpu/csr_sum.cc:60]: Add some docstring (e.g. `prefix sum`)

**Review by yzh119 — APPROVED** [review]

**Comment by yzh119** [comment]
Considering the indices is already sorted. We don't actually need a hashmap to derive non-zero elements.
For example, suppose you are computing a row of A = B + C (both B and C are coalesced sparse matrices), you can simply iterate over the sparse axis of both B and C in a ascending order:
```c++
ib = indptr_b[row];
ic = indptr_c[row];
while (ib < indptr_b[row + 1] || ic < indptr_c[row + 1]) {
  if (indices_b[ib] == indices_c[ic]) {
    // compute b_{ij} + c_{ij} and store it in corresponding position of a.
    ++ib;
    ++ic;
  } else if (indices[ib] < indices_c[ic]) {
    // store b_{ij} in corresponding position of a.
    ++ib;
  } else {
    // store c_{ij} in corresponding position of a.
    ++ic;
  }
}
```

File: src/array/cpu/csr_mm.cc (hunk 2 above)
Comment (BarclayII) [reply src/array/cpu/csr_mm.cc:70]: It should be thread-local and also initialized before entering the parallel region (unlike `private`).

https://stackoverflow.com/questions/15304760/how-are-firstprivate-and-lastprivate-different-than-private-clauses-in-openmp

File: src/array/cpu/csr_mask.cc [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  const IdType* B_indices = B.indices.Ptr<IdType>();
+  const IdType* B_eids = B_has_eid ? B.data.Ptr<IdType>() : nullptr;
+  const DType* A_data = A_weights.Ptr<DType>();
+  const int64_t M = A.num_rows;
+  const int64_t N = A.num_cols;
+
+  NDArray C_weights = NDArray::Empty({B.indices->shape[0]}, A_weights->dtype, A_weights->ctx);
+  DType* C_data = C_weights.Ptr<DType>();
+
+  phmap::flat_hash_map<IdType, DType> map;
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/cpu/csr_mask.cc:40]: Yes; I think phmap is a more-optimized drop-in replacement of `std::unordered_map`.

File: src/array/kernel.cc [hunk 5]
```diff
@@ -129,6 +129,86 @@ void BackwardSegmentCmpDispatch(NDArray feat, NDArray arg, NDArray out) {
...
+      {"A's edge weights", "B's edge weights"});
+  CHECK_EQ(A.indptr->ctx, B.indptr->ctx) << "Device of two graphs must match.";
+  CHECK_EQ(A.indptr->dtype, B.indptr->dtype) << "ID types of two graphs must match.";
+  CHECK_EQ(A_weights->dtype, B_weights->dtype) << "Data types of two edge weights must match.";
+
+  std::pair<CSRMatrix, NDArray> ret;
+  // TODO(BarclayII): change to ATEN_XPU_SWITCH_CUDA once the GPU kernels are implemented
+  ATEN_XPU_SWITCH(A.indptr->ctx.device_type, XPU, "CSRMM", {
+    ATEN_ID_TYPE_SWITCH(A.indptr->dtype, IdType, {
+      ATEN_FLOAT_TYPE_SWITCH(A_weights->dtype, DType, "Edge weights", {
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply src/array/kernel.cc:149]: K.  Just to confirm that we don't support fp16 CPU, right?

File: src/array/kernel_decl.h (hunk 1 above)
Comment (BarclayII) [reply src/array/kernel_decl.h:112]: I think "Add" implies a binary operation while "Sum" can be n-ary.  This is an n-ary operation for more efficient gradient computation.

File: python/dgl/sparse.py [hunk 6]
```diff
@@ -366,5 +367,101 @@ def _bwd_segment_cmp(feat, arg, m):
...
+    B_indptr = F.zerocopy_from_numpy(B.indptr)
+    B_indices = F.zerocopy_from_numpy(B.indices)
+    B_data = _CAPI_DGLCSRMask(
+        A.shape[0], A.shape[1],
+        F.to_dgl_nd(A_indptr),
+        F.to_dgl_nd(A_indices),
+        F.to_dgl_nd(A_data),
+        F.to_dgl_nd(B_indptr),
+        F.to_dgl_nd(B_indices))
+    return F.asnumpy(F.from_dgl_nd(B_data))
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/sparse.py]: Fixed docstring.

File: src/array/kernel.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yzh119) [reply src/array/kernel.cc:149]: yep.

File: src/array/kernel_decl.h (hunk 1 above)
Comment (yzh119) [reply src/array/kernel_decl.h:112]: Got it.

**Comment by BarclayII** [comment]
I would leave that to future PRs.  Linear scanning for mask and mm is fine, but for summing I probably need to implement a n-way merging algorithm.

**Pull request merged**